Include/internal/pycore_ceval.h

```diff
@@ -71,6 +74,11 @@ _PyEval_EvalFrame(PyThreadState *tstate, struct _PyInterpreterFrame *frame, int
 {
     EVAL_CALL_STAT_INC(EVAL_CALL_TOTAL);
     if (tstate->interp->eval_frame == NULL) {
+        PyCodeObject *co = frame->f_code;
+        py_trampoline f = (py_trampoline)(co->co_trampoline);
+        if (f) {
+            return f(_PyEval_EvalFrameDefault, tstate, frame, throwflag);
```

Comment:
rather than pass the address of `_PyEval_EvalFrameDefault` in, you can make a lighter weight trampoline: you're generating the code, inline the address of this as an immediate into the generated code. either an immediate constant load or a jump or call target - whatever the architecture and desired code requires.

that way your function signature can be identical to that of `_PyEval_EvalFrameDefault`.
With that in mind, can your trampoline skip its own call and just be an immediate jump directly to the function? shrinking the required trampoline code and saving stack space - `_PyEval_EvalFrameDefault` would then not show up in the stack at all, instead showing the name of your trampoline.  _(I don't know if that result is desirable)_

Reply:
> rather than pass the address of _PyEval_EvalFrameDefault in, you can make a lighter weight trampoline: you're generating the code, inline the address of this as an immediate into the generated code. either an immediate constant load or a jump or call target - whatever the architecture and desired code requires.

That would require relocation of the address, wouldn't it? For the first version, I prefer to not have to deal with that as there are other challenges we need to solve first. We can improve this once the feature is implemented.

Message #profiling-and-debugging

Reply:
> That would require relocation of the address, wouldn't it?

I don't think so. `&_PyEval_EvalFrameDefault` is an absolute address. I'm not aware of any important architecture being unable to jump or call to absolute addresses.

An implementation/iteration of this demonstrated internally filled in the address in the code template while populating a page of trampolines but is otherwise basically the same. (good validation of the idea if nothing else)

I'm pretty happy that we're all on the same page (pun intended) with how and why this concept works and is useful.

A jump rather than the call idea is me thinking out loud, it probably has consequences. Just something to ponder. Agreed: future.

Reply:
> I don't think so. &_PyEval_EvalFrameDefault is an absolute address. I'm not aware of any important architecture being unable to jump or call to absolute addresses.

It is, but it requires a relocation because it's address is not know until the linker starts to assemble the final binary. The compiler (or us) should place 0 in the `call` target and emit a relocation so the linker knows that needs to place the final address of `_PyEval_EvalFrameDefault`. This should be happening automatically but for some reason I'm getting segfaults :(

Reply:
Ok, found the problem with the relocation.

The problem is that when we compile the ASM, then the compiler generates one relocation for _PyEval_EvalFrameDefault of type R_X86_64_PLT32. The compiler leaves space for the address of the function in the generated assembly:

```
   0:   55                      push   rbp
   1:   48 89 e5                mov    rbp,rsp
   4:   e8 00 00 00 00          call   9 <_Py_trampoline_func_start+0x9>        5: R_X86_64_PLT32       _PyEval_EvalFrameDefault-0x4
   9:   5d                      pop    rbp
   a:   c3                      ret
```

That 00 00 00 00 will be filled by the linker with the appropiate address. But because `call` opcode takes an offset from the instruction pointer, it means that the offset that will be placed here is unique and is always relative to the location or where our template ends in memory. For example:

```
(gdb) x/11bx &_Py_trampoline_func_start
0x5555556d97e0 <_Py_trampoline_func_start>:     0x55    0x48    0x89    0xe5    0xe8    0x06    0x79    0x03
0x5555556d97e8 <_Py_trampoline_func_start+8>:   0x00    0x5d    0xc3
```

That  0x06    0x79    0x03 0x00 is the offset (0x00037906  into account). But that's an offset relative to the location of `_Py_trampoline_func_start` + offset to the call instruction:

```
info symbol (0x00037906+0x5555556d97e0+8*6)
_PyEval_EvalFrameDefault + 39 in section .text of /home/pablogsal/github/python/main/_bootstrap_python
```

As we are making multiple `mmap` places, the instruction pointer changes every time (because the address of the `mmap` segment chages and therefore the offset is wrong and should be updated to the start of our `mmap` region, which is nasty and I don't think I want to deal with this in the first version

Reply:
So for the first version, I am going to stick to the current version that takes the address of the final call as a pointer that is already relocated. We can improve it in a future PR

Reply:
thanks for the deep dive on that. agreed.  there might be possible simpler trampoline options if the pointer to the function were passed last rather than first, but that kind of thing can be sorted out as a followup if so. especially when aarch64 support is added which i expect to add a bit more stuff.

Reply:
We have support aarch64 (LE, 64bit ABI). I used `aarch64-linux-gnu-gcc` to generate ASM and qemu-static + podman to verify that it works.

Reply:
And I have implemented your suggestion to move the function pointer last. It halfed the sized of the ASM trampoline, https://github.com/pablogsal/cpython/pull/38/files